File: web/lib/publish.ts
"use server";

import { execFile } from "child_process";
import { promises as fs } from "fs";
import path from "path";
import { promisify } from "util";

import { revalidatePath } from "next/cache";

import { DATA_DIR } from "./data";
import { getEditor } from "./session";

const run = promisify(execFile);
const REPO = path.join(/* turbopackIgnore: true */ process.cwd(), "..");

/**
 * Publishing, and the check that stands in front of it.
 *
 * Unlike commissioning this is fast — grading a beat sheet is arithmetic, no
 * model call — so it runs synchronously and the reader gets an answer rather
 * than a progress page.
 *
 * The check is deliberately re-run by Python rather than reimplemented here.
 * `validate_output` is the definition of what a sound season is; a second
 * implementation in TypeScript would drift from it, and the drift would show up
 * as a season the console calls fine and the pipeline refuses.
 *
 * What is done here instead is translation. The checker writes for whoever
 * repairs the data — beat ids, field names, the two legal shapes of
 * `source_ref` — and nobody on the commissioning side has seen any of those
 * words. Every shape it can produce is said below as the thing it would tell an
 * editor, the same way the slate rewrites the loader's notes.
 */

/** One episode that has gone out, as `publish.json` recorded the decision. */
export interface EpisodeRelease {
  ep: number;
  /** Editor id, as `--by` wrote it. Null on a release recorded without one. */
  by: string | null;
  at: string | null;
}

/**
 * Two decisions, not one, because the platform earns per unlocked episode.
 *
 * `live` says the show exists for listeners at all; the episode fields say what
 * they can actually reach. A show live with nothing out is a real pre-launch
 * state, so `live: true` with an empty `episodes` is not a broken read.
 *
 * `live` / `by` / `at` keep their old meaning — they are the season decision,
 * not the newest episode one — because components already read them.
 */
export interface PublishState {
  live: boolean;
  by: string | null;
  at: string | null;
  /** Episodes written, counted off disk. Nothing to do with how many are out. */
  episodeCount: number;
  /** Every per-episode record on file, ascending. */
  episodes: EpisodeRelease[];
  /**
   * The last episode a listener can reach: the unbroken run from episode 1,
   * counted exactly as `released_through` counts it. A hole punched into the
   * file by hand reads as "out up to the hole" rather than offering an episode
   * nobody can get to.
   */
  releasedThrough: number;
  /**
   * The one episode `publish_episode` would accept next, or null when there is
   * none — not live, nothing written, or everything already out.
   *
   * This is an affordance, not a rule. Python decides what it will accept; this
   * only lets the screen avoid offering a button it knows would be refused.
   */
  nextRelease: number | null;
}

/**
 * What kind of fault a finding is.
 *
 * Only `contradiction` means two statements in the canon cannot both be true.
 * `bookkeeping` is a malformed record — a name that is not a character, a
 * moment that points at no source. Both block, and they block for different
 * reasons, so the screen must not call the second one the first: a producer
 * told the season contradicts itself concludes the writing is broken when the
 * data entry is. Same rule the validator panel keeps for `error` versus `warn`.
 */
export type FaultKind = "contradiction" | "bookkeeping" | "reading" | "unclassified";

export interface Finding {
  /** What the producer reads. Always populated. */
  said: string;
  /** The checker's own line. Rendered only when nothing could rewrite it. */
  raw: string;
  kind: FaultKind;
}

export interface Checks {
  /**
   * Null when the check ran to completion. Otherwise why it could not, in
   * words a producer can act on — and the panel refuses to publish on it. A
   * check that did not run has to read as a refusal rather than as a clean
   * season, or the guarantee is decorative exactly when the machine is broken.
   */
  unavailable: string | null;
  fatal: Finding[];
  advisory: Finding[];
}

const NOT_RUN =
  "The continuity check could not be run on this machine, so nothing is standing behind this season yet. Until it runs there is no result to trust, and an unanswered check is not a passed one.";

const NO_WORDS_FOR_IT =
  "The continuity check found something this screen has no plain words for yet. In its own words:";

/**
 * The seven generators in `validate_output`, said as an editor would hear them.
 *
 * `untraceable_beats`, `unknown_participants` and `contradictory_beats` are the
 * fatal three; `unstated_ignorance`, `present_but_unstated`, `thin_characters`
 * and `alleged_as_fact` are advisory. The empty beat sheet `validate_output`
 * appends itself is the eighth shape. Anything matching none of them is still
 * shown, under a lead-in saying so — a producer reading nothing is worse off
 * than a producer reading jargon.
 *
 * Beat ids are dropped rather than translated. `b014, b030` names nothing an
 * editor can look up, and the counts beside it carry the whole point.
 *
 * The separator is matched as ` \S+ ` rather than as a literal em dash. Python
 * logs to stderr in the console codepage on Windows, so the dash can arrive as
 * a replacement character; `PYTHONIOENCODING` below stops that at the source,
 * and matching by shape means a rewrite still fires if it ever comes back.
 */
const REWRITES: [RegExp, string, FaultKind][] = [
  [
    /^(\d+) different source_ref formats across (\d+) beats, none of them '.+?#<timeline_id>' or 'fictionalized': (.+), \.\.\.$/,
    "$2 moments do not say where they came from. Each one has to point at a line in the research or be marked as invented; instead there are $1 different labels, among them $3. Without it nobody can answer which parts of this really happened.",
    "bookkeeping",
  ],
  [
    /^source_ref '(.*)' on (\d+) beats \([^)]*\) is not '.+?#<timeline_id>' or 'fictionalized'(.*)$/,
    "$2 moments are filed under “$1”, which points at nothing: a moment has to name the line of research it came from, or be marked as invented.$3",
    "bookkeeping",
  ],
  [
    /^'(.+)' is a participant on (\d+) beats \([^)]*\) but is not a cast char_id$/,
    "“$1” is recorded as being in $2 scenes, but there is nobody by that name in the cast. Anything named in a scene becomes a character everywhere after this — with its own record of what it knows, offered as someone who could carry a spin-off.",
    "bookkeeping",
  ],
  [
    /^(\S+): \[(.+)\] are both witnessed_by and hidden_from \S+ knows and blind cannot both hold$/,
    "One scene records $2 as both having seen it and never finding out. Both cannot be true of the same person, and the spin-offs are generated from that record.",
    "contradiction",
  ],
  [
    /^the beat sheet is empty \S+ there is no canon to query$/,
    "Nothing that happens in this season has been recorded, so there is nothing to check and nothing a spin-off could be built from.",
    "bookkeeping",
  ],
  [
    /^(\S+) \(ep(.+?)\) asserts nobody's ignorance: hidden_from is empty and (\d+) of (\d+) cast members are not even present$/,
    "A moment in episode $2 records nobody as being kept in the dark by it, though $3 of the $4 characters are not even in the scene. Right for something that happens in public, a hole anywhere else — only reading it settles which.",
    "reading",
  ],
  [
    /^(\d+) of (\d+) beats leave someone standing in the scene with no stated knowledge of it \S+ present, but in neither witnessed_by nor hidden_from \(.*\)$/,
    "$1 of $2 moments put someone in the scene without saying whether they take it in. Anyone left like that counts as not knowing, which is how a spin-off lead ends up blind to their own scenes.",
    "reading",
  ],
  [
    /^(\d+) cast members appear in no beat at all \(([^)]*)\) \S+ nothing to promote, and the character panel would offer them anyway$/,
    "Cast with no scenes at all: $2. There is nothing there to build a spin-off from, and the character list would offer them anyway.",
    "reading",
  ],
  [
    /^(\d+) of (\d+) cast members have their status stated on fewer than half of (\d+) beats \S+ (.+)\. A spinoff lead needs this at zero unstated$/,
    "For $1 of $2 characters, whether they know a thing or not is recorded on fewer than half of the season’s $3 moments — $4. Any of them could be picked to carry a spin-off, and a lead needs it recorded everywhere.",
    "reading",
  ],
  [
    /^(\S+) is sourced to (\S+) \((alleged|disputed)\) \S+ hard rule 3 allows it only as something a character claims: "(.*)"$/,
    "A moment is drawn from something the record only has as $3, so it can reach the script as an accusation a character makes and never as something the show states: “$4”",
    "reading",
  ],
];

/**
 * Second pass. The repair hints `untraceable_beats` tacks onto a sentence, and
 * the counts it writes in passing, are their own small phrases rather than
 * whole messages — so they are rewritten after the sentence they hang off.
 */
const TIDY: [RegExp, string][] = [
  [
    /\. Timeline entry '(.+?)' exists; write it as '.+?'/,
    " The research does have a line called “$1”; this points at it wrongly.",
  ],
  [
    /\. Right entry, wrong event id: this dossier is '.+?'/,
    " It names the right line of research, but under the wrong story.",
  ],
  [
    /\. Invention is marked with the literal 'fictionalized'/,
    " It looks like invention that was marked in the wrong words.",
  ],
  // `(missing)` is the checker's stand-in for an empty field, not a label
  // anybody typed, so it is not quoted back as though it were one.
  [
    /filed under “\(missing\)”, which points at nothing: a/,
    "filed under nothing at all. A",
  ],
  [/\((\d+) unstated\)/g, "($1 unrecorded)"],
  // The checker writes "1 beats" throughout; a screen an editor reads cannot.
  [/\bin 1 scenes\b/, "in one scene"],
  [/^1 moments are filed under/, "One moment is filed under"],
  [/^1 of (\d+) moments put someone/, "One of $1 moments puts someone"],
  [/\b1 of the (\d+) characters are not\b/, "1 of the $1 characters is not"],
];

/** One checker line, said plainly. Never drops it — the worst case is verbatim. */
function plainly(raw: string): Finding {
  for (const [pattern, says, kind] of REWRITES) {
    if (!pattern.test(raw)) continue;
    let said = raw.replace(pattern, says);
    for (const [from, to] of TIDY) said = said.replace(from, to);
    return { said, raw, kind };
  }
  return { said: NO_WORDS_FOR_IT, raw, kind: "unclassified" };
}

/**
 * The checker's own last line: `story: N fatal, M advisory`.
 *
 * It is the only proof the check actually ran. Nothing else distinguishes a
 * clean season from a Python that could not be started, a season directory that
 * does not exist, or a run killed by the timeout — all three produce no ERROR
 * and no WARN lines, which read as a pass.
 */
const SUMMARY = /INFO\s+.*: (\d+) fatal, (\d+) advisory/;

/**
 * `^ep(\d+)\.md$`, deliberately case-sensitive.
 *
 * `episode_count` in src/publish.py matches exactly this, and the count is what
 * Python compares an episode number against before it will release one. A
 * looser pattern here would offer an episode 15 that the backend then refuses,
 * which is the one thing the disabled-control rule is meant to avoid. Note this
 * is stricter than `serials.ts`, which matches case-insensitively.
 */
const EP_FILE = /^ep(\d+)\.md$/;

/**
 * Counted off disk only when the caller has no count already.
 *
 * A page rendering a season has loaded it, and `Serial.episodeCount` is the
 * same number; passing it in spends one fewer directory read per render. The
 * fallback exists so this function is still correct when called on its own —
 * `serials.ts` is not editable from here and exports no cheap counter.
 */
async function countEpisodes(storyId: string): Promise<number> {
  try {
    const names = await fs.readdir(
      path.join(DATA_DIR, "stories", storyId, "episodes"),
    );
    return names.filter((n) => EP_FILE.test(n)).length;
  } catch {
    // No `episodes/` at all: planned, never written. Zero, not an error.
    return 0;
  }
}

/**
 * The `episodes` map, read the way `released()` reads it.
 *
 * Absent entirely on every file written before episodes had their own state, so
 * a missing key is "none out" rather than a fault. A key that is not an episode
 * number is somebody's hand edit and is skipped alone — losing every recorded
 * decision to one bad key is the worse failure.
 */
function releases(raw: unknown): EpisodeRelease[] {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return [];
  const out: EpisodeRelease[] = [];
  for (const [key, value] of Object.entries(raw as Record<string, unknown>)) {
    const ep = Number(key);
    if (!Number.isInteger(ep) || ep < 1) continue;
    const r =
      value && typeof value === "object" ? (value as Record<string, unknown>) : {};
    out.push({
      ep,
      by: typeof r.by === "string" ? r.by : null,
      at: typeof r.at === "string" ? r.at : null,
    });
  }
  return out.sort((a, b) => a.ep - b.ep);
}

/** The unbroken run from episode 1. Mirrors `released_through`. */
function unbrokenRun(eps: EpisodeRelease[]): number {
  const out = new Set(eps.map((e) => e.ep));
  let n = 0;
  while (out.has(n + 1)) n += 1;
  return n;
}

/**
 * `episodeCount` is optional: pass `Serial.episodeCount` where the season is
 * already loaded, and this reads one file instead of a file and a directory.
 */
export async function readPublishState(
  storyId: string,
  episodeCount?: number,
): Promise<PublishState> {
  const total =
    typeof episodeCount === "number" && Number.isFinite(episodeCount) && episodeCount >= 0
      ? Math.floor(episodeCount)
      : await countEpisodes(storyId);

  // A season is a draft until somebody says so, and a draft has nothing out
  // whatever is written. Also where an unreadable file lands: a state file that
  // will not parse must not read as a live show.
  const draft: PublishState = {
    live: false,
    by: null,
    at: null,
    episodeCount: total,
    episodes: [],
    releasedThrough: 0,
    nextRelease: null,
  };

  let r: Record<string, unknown>;
  try {
    const raw = JSON.parse(
      await fs.readFile(
        path.join(DATA_DIR, "stories", storyId, "publish.json"),
        "utf-8",
      ),
    );
    r = raw && typeof raw === "object" ? (raw as Record<string, unknown>) : {};
  } catch {
    return draft;
  }

  if (r.state !== "live") return draft;

  const episodes = releases(r.episodes);
  const releasedThrough = unbrokenRun(episodes);

  return {
    live: true,
    by: typeof r.by === "string" ? r.by : null,
    at: typeof r.at === "string" ? r.at : null,
    episodeCount: total,
    episodes,
    releasedThrough,
    // Episodes go out in order, so there is only ever one candidate. Null once
    // the season is fully out, and null when nothing is written to release.
    nextRelease: releasedThrough < total ? releasedThrough + 1 : null,
  };
}

/** Reads the checker's log. Exit code 1 means fatal problems, not a crash. */
export async function readChecks(storyId: string): Promise<Checks> {
  let output = "";
  try {
    const { stdout, stderr } = await run(
      "python",
      ["-m", "src.publish", "--story", storyId, "--check"],
      {
        cwd: REPO,
        timeout: 30_000,
        // The checker writes em dashes. Without this Python encodes stderr in
        // the console codepage on Windows and Node decodes it as UTF-8, so
        // every dash reaches the screen as a replacement character.
        env: { ...process.env, PYTHONIOENCODING: "utf-8" },
      },
    );
    output = `${stdout}\n${stderr}`;
  } catch (err) {
    // A rejection is the ordinary case: exit 1 means fatal problems were found
    // and the log is on stderr. It is also what an unresolvable `python` looks
    // like — and promisified execFile attaches empty strings rather than
    // nothing for ENOENT, so a failure to start cannot be told apart here. The
    // summary line below is what tells them apart.
    const e = err as { stdout?: string; stderr?: string };
    output = `${e.stdout ?? ""}\n${e.stderr ?? ""}`;
  }

  const summary = output.match(SUMMARY);
  if (!summary) return { unavailable: NOT_RUN, fatal: [], advisory: [] };

  const fatal: Finding[] = [];
  const advisory: Finding[] = [];
  for (const line of output.split(/\r?\n/)) {
    const isFatal = line.match(/ERROR\s+FATAL\s+(.+)$/);
    if (isFatal) fatal.push(plainly(isFatal[1].trim()));
    const isAdvisory = line.match(/WARN\s+advisory\s+(.+)$/);
    if (isAdvisory) advisory.push(plainly(isAdvisory[1].trim()));
  }

  // The checker counts its own findings. If fewer lines were scraped than it
  // says it wrote, the difference is said rather than lost — a count that
  // quietly shrinks on the way to the screen is the same failure as a check
  // that quietly passes.
  const short = Number(summary[1]) - fatal.length;
  if (short > 0) {
    fatal.push({
      said: `The check reported ${short} further problem${short === 1 ? "" : "s"} that this screen could not read. Treat the season as unsound until someone has looked at the check itself.`,
      raw: "",
      kind: "unclassified",
    });
  }

  return { unavailable: null, fatal, advisory };
}

/**
 * Every screen a release decision moves.
 *
 * The episode pages are refreshed by route pattern rather than one literal
 * path, because no release decision touches a single episode: pulling episode 3
 * pulls everything after it, and taking the season back to draft pulls all of
 * them. Refreshing only the episode named on the form would leave the tail
 * showing itself as out.
 */
function revalidateSeason(storyId: string): void {
  revalidatePath(`/serials/${storyId}`);
  revalidatePath("/serials");
  revalidatePath("/serials/[id]/[ep]", "page");
}

/** The `ep` field off a form, or null if it is not an episode number. */
function episodeArg(formData: FormData): number | null {
  const n = Number(String(formData.get("ep") ?? "").trim());
  return Number.isInteger(n) && n >= 1 ? n : null;
}

export async function publishSeason(formData: FormData): Promise<void> {
  const storyId = String(formData.get("storyId") ?? "").trim();
  if (!storyId) return;

  const editor = await getEditor();
  const args = ["-m", "src.publish", "--story", storyId];
  if (editor) args.push("--by", editor.id);

  try {
    await run("python", args, { cwd: REPO, timeout: 60_000 });
  } catch {
    // A refusal is not an error page. The screen already shows the same checks
    // this would report, so re-rendering says why on its own. `publish()`
    // re-runs them before it writes, so a request that arrives without the
    // button being shown is refused there rather than trusted here.
  }
  revalidateSeason(storyId);
}

export async function unpublishSeason(formData: FormData): Promise<void> {
  const storyId = String(formData.get("storyId") ?? "").trim();
  if (!storyId) return;
  try {
    await run("python", ["-m", "src.publish", "--story", storyId, "--unpublish"], {
      cwd: REPO,
      timeout: 30_000,
    });
  } catch {
    // Pulling something back is never gated, so a failure here is a broken
    // install rather than a refusal. The state file is the truth either way.
  }
  revalidateSeason(storyId);
}

/**
 * Put one episode in front of listeners.
 *
 * Same timeout as the season: `publish_episode` re-runs the season's fatal
 * checks on every release, not only the first, so this is as slow as the first
 * publish and for the same reason.
 *
 * Three things can refuse it — the show is not live, the episode before it is
 * not out, the season's checks now fail — and all three are refused in Python.
 * None of them is decided again here. The screen re-renders against the state
 * file and the checks afterwards, which is what says why.
 */
export async function publishEpisode(formData: FormData): Promise<void> {
  const storyId = String(formData.get("storyId") ?? "").trim();
  const ep = episodeArg(formData);
  if (!storyId || ep === null) return;

  const editor = await getEditor();
  const args = [
    "-m",
    "src.publish",
    "--story",
    storyId,
    "--episode",
    String(ep),
  ];
  if (editor) args.push("--by", editor.id);

  try {
    await run("python", args, { cwd: REPO, timeout: 60_000 });
  } catch {
    // Refusals arrive here as exit 1. Not an error page, for the same reason
    // publishSeason's are not: the next render reads the state file and the
    // checks, and says what actually stands in the way.
  }
  revalidateSeason(storyId);
}

/**
 * Pull one episode, and with it every episode after it.
 *
 * The tail comes too — that is `unpublish_episode`'s rule, not this one's, and
 * it is why the screen must say so before the click rather than reimplement it
 * after. Pulling is never gated, so a failure here is a broken install.
 */
export async function unpublishEpisode(formData: FormData): Promise<void> {
  const storyId = String(formData.get("storyId") ?? "").trim();
  const ep = episodeArg(formData);
  if (!storyId || ep === null) return;

  try {
    await run(
      "python",
      ["-m", "src.publish", "--story", storyId, "--episode", String(ep), "--unpublish"],
      { cwd: REPO, timeout: 30_000 },
    );
  } catch {
    // Nothing gates a pull, so the state file is the truth either way.
  }
  revalidateSeason(storyId);
}

File: src/publish.py
"""
Put a season in front of listeners, or refuse to.

    python tasks.py publish --story a3c321dba0a7 --by priya
    python tasks.py publish --story a3c321dba0a7 --episode 3 --by priya

Writing a season and deciding it can go out are different acts, and only the
second one is a person's. Until this existed there was no difference on disk
between what the machine produced and what an editor had stood behind.

The refusal is the point. A season whose beat sheet has fatal problems cannot be
published — not by an editor, not by anyone — for the same reason a `blocked`
story cannot be commissioned: continuity is the thing this product sells, and a
guarantee that can be waived under deadline is not a guarantee. Advisories are
shown and do not block, because they are the ones only a human reading the prose
can settle.

Two decisions, not one, because the platform earns per unlocked episode. A show
going live means it exists for listeners at all; an episode going out is the
thing that actually earns. A show can be live with nothing released yet — that
is a real pre-launch state, not a broken one.

Episodes release in order and are pulled from the end. A serial with episode 7
out and episode 3 held is not a state anyone can listen to, so it is not a state
this can be put into: releasing requires the one before it to be out already,
and pulling one pulls everything after it.

There is no Pocket FM API to push to. This records the decision and flips the
state; it does not put anything in an app, and the console says so rather than
implying otherwise.
"""

import re
import sys
import json
import pathlib
import argparse
import datetime as dt
from typing import Any, Dict, List, Optional, Sequence, Tuple

from src.util import DATA, log
from src.scoring.validate import load_story, validate_output

STORIES = DATA / "stories"

EPISODE_FILE = re.compile(r"^ep(\d+)\.md$")


def _now() -> str:
    return dt.datetime.now(dt.timezone.utc).isoformat()


def _write_state(story_id: str, state: Dict[str, Any]) -> Dict[str, Any]:
    path = state_path(story_id)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(state, ensure_ascii=False, indent=2), encoding="utf-8")
    return state


def state_path(story_id: str) -> pathlib.Path:
    return STORIES / story_id / "publish.json"


def read_state(story_id: str) -> Optional[Dict[str, Any]]:
    path = state_path(story_id)
    if not path.exists():
        return None
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None


def episode_numbers(story_id: str) -> set:
    """
    Which episodes exist on disk, by number.

    The numbers rather than the count, because the two disagree the moment a
    season has a hole in it — and a count would then refuse to release an
    episode whose script is sitting right there.
    """
    episodes = STORIES / story_id / "episodes"
    if not episodes.is_dir():
        return set()
    found = set()
    for p in episodes.iterdir():
        match = EPISODE_FILE.match(p.name) if p.is_file() else None
        if match:
            found.add(int(match.group(1)))
    return found


def episode_count(story_id: str) -> int:
    """How many episodes were actually written, counted off disk."""
    return len(episode_numbers(story_id))


def released(state: Optional[Dict[str, Any]]) -> Dict[int, Dict[str, Any]]:
    """The per-episode records, keyed by episode number rather than by string."""
    return _split_episodes(state)[0]


def _split_episodes(
    state: Optional[Dict[str, Any]],
) -> Tuple[Dict[int, Dict[str, Any]], Dict[str, Any]]:
    """
    The episode records, and anything else that was in there.

    A key that is not an episode number is somebody's hand edit. It is skipped
    rather than failing the whole read — the file is a record of decisions, and
    losing all of them to one bad key is worse. It is also carried back out to
    the writer, because a read that tolerates something and a write that then
    deletes it is not tolerance, only a delayed loss.
    """
    if not state:
        return {}, {}
    raw = state.get("episodes")
    if not isinstance(raw, dict):
        return {}, {}
    out: Dict[int, Dict[str, Any]] = {}
    extra: Dict[str, Any] = {}
    for key, value in raw.items():
        try:
            out[int(key)] = value if isinstance(value, dict) else {}
        except (TypeError, ValueError):
            extra[str(key)] = value
    return out, extra


def _episodes_field(
    episodes: Dict[int, Dict[str, Any]], extra: Optional[Dict[str, Any]] = None
) -> Dict[str, Any]:
    """Episode records back into their on-disk shape, keeping what we don't parse."""
    field: Dict[str, Any] = {str(k): v for k, v in sorted(episodes.items())}
    field.update(extra or {})
    return field


def released_through(state: Optional[Dict[str, Any]]) -> int:
    """
    The last episode a listener can reach.

    Counts the unbroken run from episode 1, so a hole punched into the file by
    hand reads as "released up to the hole" rather than silently offering an
    episode nobody can get to.
    """
    out = released(state)
    n = 0
    while (n + 1) in out:
        n += 1
    return n


def check(story_id: str) -> Tuple[List[str], List[str]]:
    """Grade a season on disk. `(fatal, advisory)`, same contract as the writer's."""
    story_dir = STORIES / story_id
    if not story_dir.is_dir():
        raise RuntimeError(f"no season at {story_dir}")
    dossier, beats = load_story(story_dir)
    return validate_output(dossier, beats)


def publish(story_id: str, by: Optional[str] = None) -> Dict[str, Any]:
    """
    Publish, or raise saying why not.

    The check runs against what is on disk right now rather than trusting a
    result recorded when the season was written — a season can be edited by hand
    afterwards, and this is the last gate before listeners.
    """
    fatal, advisory = check(story_id)

    if fatal:
        for note in fatal:
            log(f"FATAL {note}", "error")
        raise RuntimeError(
            f"{story_id} cannot be published: {len(fatal)} continuity "
            f"problem{'' if len(fatal) == 1 else 's'}. {fatal[0]}"
        )

    for note in advisory:
        log(f"advisory {note}", "warn")

    # Episodes already out survive a re-publish. Standing behind the season a
    # second time is not a reason to pull what listeners already have.
    existing, extra = _split_episodes(read_state(story_id))

    # ...but only the ones that still exist. Re-commissioning a 14-episode
    # season as a 3-episode taster makes `persist()` delete the surplus scripts,
    # and a release record outliving its script reported "14 of 3 episodes out"
    # while offering listeners episodes whose files were gone. Dropped here
    # rather than in `released()`, because the record should only be discarded
    # at the moment someone re-publishes and the season on disk is authoritative.
    written = episode_numbers(story_id)
    orphaned = sorted(ep for ep in existing if ep not in written)
    for ep in orphaned:
        existing.pop(ep, None)
    if orphaned:
        log(f"{story_id}: {len(orphaned)} released episode"
            f"{'' if len(orphaned) == 1 else 's'} no longer written "
            f"({', '.join(str(e) for e in orphaned)}) — dropped from what is out",
            "warn")

    state = {
        "story_id": story_id,
        "state": "live",
        "by": by,
        "at": _now(),
        # Kept with the decision: what was known about the season at the moment
        # someone stood behind it.
        "advisory_at_publish": advisory,
        "episodes": _episodes_field(existing, extra),
    }
    _write_state(story_id, state)
    log(f"{story_id} is live" + (f", published by {by}" if by else ""))
    return state


def publish_episode(story_id: str, ep: int, by: Optional[str] = None) -> Dict[str, Any]:
    """
    Put one episode in front of listeners, or raise saying why not.

    Re-runs the season's checks rather than trusting the verdict recorded when
    the show went live: episodes go out days apart, and the beat sheet can be
    edited in between. The gate that stops a broken season shipping has to stop
    a broken episode too, or it is only a gate on the first one.
    """
    state = read_state(story_id)
    if not state or state.get("state") != "live":
        raise RuntimeError(
            f"{story_id} is not live, so episode {ep} cannot go out. "
            f"Publish the show first."
        )

    written = episode_numbers(story_id)
    if not written:
        raise RuntimeError(f"{story_id} has no written episodes")
    if ep not in written:
        total = len(written)
        raise RuntimeError(
            f"{story_id} has {total} episode{'' if total == 1 else 's'} written; "
            f"there is no episode {ep}"
        )

    out, extra = _split_episodes(state)
    if ep in out:
        return state

    if ep > 1 and (ep - 1) not in out:
        raise RuntimeError(
            f"episode {ep - 1} is not out yet, so episode {ep} cannot be. "
            f"A serial is listened to in order."
        )

    fatal, advisory = check(story_id)
    if fatal:
        for note in fatal:
            log(f"FATAL {note}", "error")
        raise RuntimeError(
            f"episode {ep} of {story_id} cannot go out: {len(fatal)} continuity "
            f"problem{'' if len(fatal) == 1 else 's'} in the season. {fatal[0]}"
        )

    # The check is re-run here to catch drift since launch, so throwing away
    # what it found would defeat the point of running it. The editor releasing
    # this episode is the only person in a position to settle an advisory, and
    # the season-level list was written before these ones existed.
    for note in advisory:
        log(f"advisory {note}", "warn")

    out[ep] = {"by": by, "at": _now(), "advisory_at_release": advisory}
    state["episodes"] = _episodes_field(out, extra)
    _write_state(story_id, state)
    log(f"{story_id} episode {ep} is out" + (f", released by {by}" if by else ""))
    return state


def unpublish_episode(story_id: str, ep: int) -> Dict[str, Any]:
    """
    Pull one episode, and everything after it.

    Pulling is never gated — only shipping is. But pulling episode 3 of a season
    that has 7 out would leave a hole a listener cannot get past, so the tail
    comes with it.
    """
    state = read_state(story_id)
    if not state:
        raise RuntimeError(f"{story_id} is not live")

    out, extra = _split_episodes(state)
    pulled = sorted(n for n in out if n >= ep)
    for n in pulled:
        out.pop(n, None)

    state["episodes"] = _episodes_field(out, extra)
    _write_state(story_id, state)
    if pulled:
        tail = "" if len(pulled) == 1 else f" (and {len(pulled) - 1} after it)"
        log(f"{story_id} episode {ep} pulled{tail}")
    return state


def unpublish(story_id: str) -> None:
    """
    Back to draft, episodes and all. Pulling is never gated — only shipping is.
    """
    path = state_path(story_id)
    if path.exists():
        path.unlink()
        log(f"{story_id} is back to draft")


def main(argv: Optional[Sequence[str]] = None) -> int:
    parser = argparse.ArgumentParser(
        prog="tasks.py publish",
        description="Publish a written season, if its continuity checks pass.",
    )
    parser.add_argument("--story", required=True, help="directory under data/stories")
    parser.add_argument("--by", default=None, help="who is publishing it")
    parser.add_argument("--check", action="store_true",
                        help="report the checks without publishing")
    parser.add_argument("--unpublish", action="store_true", help="return it to draft")
    parser.add_argument("--episode", type=int, default=None, metavar="N",
                        help="release one episode; with --unpublish, pull it "
                             "and everything after it")
    parser.add_argument("--status", action="store_true",
                        help="what is live and how many episodes are out")
    args = parser.parse_args(argv)

    try:
        if args.status:
            state = read_state(args.story)
            total = episode_count(args.story)
            if not state or state.get("state") != "live":
                log(f"{args.story}: not live, {total} episodes written")
                return 0
            through = released_through(state)
            log(f"{args.story}: live, {through} of {total} episodes out")
            for ep, rec in sorted(released(state).items()):
                log(f"  ep{ep:02d} out"
                    + (f" — {rec.get('by')}" if rec.get("by") else ""))
            return 0

        if args.unpublish:
            if args.episode is not None:
                unpublish_episode(args.story, args.episode)
            else:
                unpublish(args.story)
            return 0

        if args.check:
            fatal, advisory = check(args.story)
            for note in fatal:
                log(f"FATAL {note}", "error")
            for note in advisory:
                log(f"advisory {note}", "warn")
            log(f"{args.story}: {len(fatal)} fatal, {len(advisory)} advisory")
            return 1 if fatal else 0

        if args.episode is not None:
            publish_episode(args.story, args.episode, args.by)
            return 0

        publish(args.story, args.by)
    except (RuntimeError, ValueError, KeyError, json.JSONDecodeError) as exc:
        log(str(exc), "error")
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
